core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java

```diff
@@ -546,26 +560,33 @@ public int register(String subject,
       }
 
       // determine the latest version of the schema in the subject
-      List<SchemaValue> allVersions = getAllSchemaValues(subject);
+      List<SchemaKey> allVersions = getAllSchemaKeys(subject);
       Collections.reverse(allVersions);
 
       List<SchemaValue> deletedVersions = new ArrayList<>();
-      List<ParsedSchema> undeletedVersions = new ArrayList<>();
+      List<ParsedSchemaHolder> undeletedVersions = new ArrayList<>();
       int newVersion = MIN_VERSION;
-      for (SchemaValue schemaValue : allVersions) {
```

Comment:
Just for my understanding, this is the root cause for the OOM since all the schema versions are referenced and can only be potentially GC'ed after this method. Now `SchemaKey` is used which itself has a smaller memory footprint and `SchemaValue`s can be GC'ed even in middle of the call due to the use of `SoftReference`. Is this correct?

Reply:
Yes you are correct.  Plus we explicitly `schemaHolder.clear` to release references to the `SchemaValue` instances.